File: generated_code/python/solution_1895_3.py
from typing import List

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        """
        Algorithm: Linear Scan (Prefix and Suffix passes)
        Time Complexity: O(n) - Two passes over the array of length n.
        Space Complexity: O(1) auxiliary space (excluding the output array).
        
        Instead of a naive O(n^2) approach, we can track the cumulative operations 
        needed from left to right, and then right to left.
        
        As we move from one box to the next, every ball we have seen so far requires 
        exactly 1 additional step.
        """
        n = len(boxes)
        ans = [0] * n
        
        # Pass 1: Left to Right
        # Track the number of balls seen so far and the operations accumulated
        balls_count = 0
        ops = 0
        for i in range(n):
            ans[i] += ops
            if boxes[i] == '1':
                balls_count += 1
            ops += balls_count
            
        # Pass 2: Right to Left
        # Accumulate the operations from the right side
        balls_count = 0
        ops = 0
        for i in range(n - 1, -1, -1):
            ans[i] += ops
            if boxes[i] == '1':
                balls_count += 1
            ops += balls_count
            
        return ans

File: generated_code/cpp/solution_1895_3.cpp
#include <vector>
#include <string>
#include <cmath>

using namespace std;

class Solution {
public:
    vector<int> minOperations(string boxes) {
        int n = boxes.length();
        vector<int> ans(n, 0);
        
        // This problem can be solved optimally in O(N) time and O(1) extra space
        // by accumulating operations from left-to-right and right-to-left.
        
        // Pass 1: Calculate operations needed to bring all balls from the left side of i to box i
        int left_balls = 0;
        int left_ops = 0;
        for (int i = 0; i < n; ++i) {
            ans[i] += left_ops;
            if (boxes[i] == '1') {
                left_balls++;
            }
            // As we move to the next box, every accumulated ball requires 1 extra operation
            left_ops += left_balls;
        }
        
        // Pass 2: Calculate operations needed to bring all balls from the right side of i to box i
        int right_balls = 0;
        int right_ops = 0;
        for (int i = n - 1; i >= 0; --i) {
            ans[i] += right_ops;
            if (boxes[i] == '1') {
                right_balls++;
            }
            // As we move to the previous box, every accumulated ball requires 1 extra operation
            right_ops += right_balls;
        }
        
        return ans;
    }
};